File: old/recDP/07-numWaysChange/my.cpp
#include <vector>
using namespace std;

int numberOfWaysToMakeChange(int n, vector<int> denoms) {
  // Write your code here.
	if (n < 0)
		return 0; // no ways of making change
	if (n == 0)
		return 1;
	if (denoms.size() == 0)
		return 0;
	vector<int> newVec(denoms.begin(), denoms.end() - 1);
	return numberOfWaysToMakeChange(n - denoms[denoms.size() - 1], denoms) + numberOfWaysToMakeChange(n, newVec);

}
//-----------

File: old/recDP/07-numWaysChange/my.py
def numberOfWaysToMakeChange(n, denoms):
    # Write your code here.
	if n < 0 :
		return 0
	if n == 0:
		return 1 # combinatorially correct
	if len(denoms) == 0:
		return 0
	# take it or leave it
	return numberOfWaysToMakeChange(n - denoms[0], denoms) + numberOfWaysToMakeChange(n, denoms[1:])



def numberOfWaysToMakeChange(n, denoms, cache = dict()):
    # Write your code here.
    if (n, len(denoms)) in cache:
        return cache[(n, len(denoms))]
    if n < 0 :
        return 0
    if n == 0:
        return 1 # combinatorially correct
    if len(denoms) == 0:
        return 0
    # take it or leave it
    cache[(n, len(denoms))] = numberOfWaysToMakeChange(n - denoms[0], denoms, cache) + numberOfWaysToMakeChange(n, denoms[1:], cache)
    return cache[(n, len(denoms))]

print(numberOfWaysToMakeChange(7, [2, 4]) == 0) # True
